src/libraries/SwapMath.sol

```diff
@@ -49,63 +51,54 @@ library SwapMath {
         unchecked {
+            uint256 _feePips = feePips; // cast once and cache
             bool zeroForOne = sqrtPriceCurrentX96 >= sqrtPriceTargetX96;
             bool exactIn = amountRemaining < 0;
 
             if (exactIn) {
-                uint256 amountRemainingLessFee = FullMath.mulDiv(uint256(-amountRemaining), 1e6 - feePips, 1e6);
+                uint256 amountRemainingLessFee =
+                    FullMath.mulDiv(uint256(-amountRemaining), MAX_FEE_PIPS - _feePips, MAX_FEE_PIPS);
                 amountIn = zeroForOne
                     ? SqrtPriceMath.getAmount0Delta(sqrtPriceTargetX96, sqrtPriceCurrentX96, liquidity, true)
                     : SqrtPriceMath.getAmount1Delta(sqrtPriceCurrentX96, sqrtPriceTargetX96, liquidity, true);
                 if (amountRemainingLessFee >= amountIn) {
+                    // `amountIn` is capped by the target price
                     sqrtPriceNextX96 = sqrtPriceTargetX96;
+                    feeAmount = _feePips == MAX_FEE_PIPS
+                        ? amountIn
+                        : FullMath.mulDivRoundingUp(amountIn, _feePips, MAX_FEE_PIPS - _feePips);
                 } else {
                     sqrtPriceNextX96 = SqrtPriceMath.getNextSqrtPriceFromInput(
                         sqrtPriceCurrentX96, liquidity, amountRemainingLessFee, zeroForOne
                     );
+                    amountIn = zeroForOne
+                        ? SqrtPriceMath.getAmount0Delta(sqrtPriceNextX96, sqrtPriceCurrentX96, liquidity, true)
+                        : SqrtPriceMath.getAmount1Delta(sqrtPriceCurrentX96, sqrtPriceNextX96, liquidity, true);
+                    // we didn't reach the target, so take the remainder of the maximum input as fee
+                    feeAmount = uint256(-amountRemaining) - amountIn;
                 }
+                amountOut = zeroForOne
+                    ? SqrtPriceMath.getAmount1Delta(sqrtPriceNextX96, sqrtPriceCurrentX96, liquidity, false)
+                    : SqrtPriceMath.getAmount0Delta(sqrtPriceCurrentX96, sqrtPriceNextX96, liquidity, false);
             } else {
                 amountOut = zeroForOne
                     ? SqrtPriceMath.getAmount1Delta(sqrtPriceTargetX96, sqrtPriceCurrentX96, liquidity, false)
                     : SqrtPriceMath.getAmount0Delta(sqrtPriceCurrentX96, sqrtPriceTargetX96, liquidity, false);
                 if (uint256(amountRemaining) >= amountOut) {
+                    // `amountOut` is capped by the target price
                     sqrtPriceNextX96 = sqrtPriceTargetX96;
                 } else {
+                    // cap the output amount to not exceed the remaining output amount
+                    amountOut = uint256(amountRemaining);
                     sqrtPriceNextX96 = SqrtPriceMath.getNextSqrtPriceFromOutput(
                         sqrtPriceCurrentX96, liquidity, uint256(amountRemaining), zeroForOne
```

Comment:
is it cheaper to pass in `amountOut` to prevent double-casting `amountRemaining`?

Reply:
Casting between `uint256` and `int256` shouldn't matter.

Reply:
was gonna ask same thing - even if it doesn't make a difference I think the code reads better with using the variable we just defined above imo

Reply:
Passing `amountOut` to `getNextSqrtPriceFromOutput`.